### ISSUE TITLE
Unsafe eval causes strict csp contexts to be served es5 bundles (then error due to unsafe eval in that compilation)

### ISSUE THREAD

**Issue by zakhenry** [stephanrauh/ngx-extended-pdf-viewer#740]
**Describe the bug**
In the following function:

https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/1335da4c2d81b1ef9ce1f9e44e9c177a19efb4e7/projects/ngx-extended-pdf-viewer/src/lib/ngx-extended-pdf-viewer.component.ts#L614-L621

The unsafe `Function()` call is made to determine whether or not a browser supports the null safe navigator, however when run with a content security policy that bans unsafe-eval, this method throws an error (but not the expected one), which means that it incorrectly assumes that the browser doesn't support null safe navigation, therefore loading the es5 implementation. The es5 implementation also has unsafe usage of `Function` (in `Function("r", "regeneratorRuntime = r")(runtime);` ) causing browsers to fail to run the worker.

**Version info**
- 9.0.0-alpha.4

**Desktop (please complete the following information):**
 - Chrome, but this will affect all evergreen browsers

**To Reproduce**
This isn't easy to reproduce unfortunately as you need to be serving from a server that can set CSP headers

**Additional context**
Relates to #687

**Comment by stephanrauh** [comment]
Oops! I thought using `Function()` is the solution to this nasty problem. Is that wrong, or did I implement it incorrectly?

**Comment by zakhenry** [comment]
The problem is that `Function` is pretty much the same as `eval`, because it is capable of executing arbitrary code.

As to how to to fix it properly, I'm really not sure as syntax errors are not catchable. It is possible that a solution might be to assume that a browser that is failing CSP is compliant? I'm not sure how true that is, as CSP has been around for a while. Maybe emitting es6 js without null safe navigation would be an alternative?

**Comment by zakhenry** [comment]
My suggestion would be to change the target in the generation of the upstream lib to es2017, this is the default that Typescript uses in their playground so it must be a good middleground for browsers?

https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/main/updateMozillasPdfViewer.sh#L30

**Comment by zakhenry** [comment]
Oh wait I misunderstood that is the bleeding edge branch, and not the core compilation. Where is that set?

**Comment by zakhenry** [comment]
Ah there it is https://github.com/mozilla/pdf.js/blob/master/gulpfile.js#L1370 hrm that is problematic

**Comment by stephanrauh** [comment]
I'm maintaining my own fork, so I don't insist on using ES2020. We can try ES2017 and see if that helps.

BTW, I intend to make version 2.8 the default branch with version 9.0.0. That's why it's still an alpha version. The problem ist, version 2.8 breaks form support, and I don't want to publish a version without that.

**Comment by zakhenry** [comment]
Yep I think using ES2017 *and* dropping the check for null safe navigation will do the trick

**Comment by stephanrauh** [comment]
I've tried to do that yesterday. But I have now idea how to do this. Adding a `.browserlists.rc` file changed almost nothing. Any suggestions?

BTW: the line you've found only modifies the TypeScript type definitions that are automatically generated. It looked like a hot trace, but to my disappointment, it isn't. :(

**Comment by zakhenry** [comment]
Ah right I see, I forgot people still write JS 😆 

I wonder if it is safest to just rely on the `Promise.allSettled` check that is already there to detect ES2020 compatibility, given it is [in the ES2020 feature set along with optional chaining](https://auth0.com/blog/javascript-whats-new-es2020/), therefore a browser that is ES2020 compliant must have both features?

So then the fix would be to simply remove the unsafe `ngxBrowserSupportsNullSafeChaining` check entirely

**Comment by stephanrauh** [comment]
Now, that's a good idea! It looks like an almost perfect match. A few Opera and Samsung users are going to experience problems, but if I did my math correctly, this affects roughly 0.4% of all users. That's still annoying, but I guess I can get away with it. Plus, I can remove a polyfill. :) Thanks for the hint!

![image](https://user-images.githubusercontent.com/3045767/116824823-85977980-ab8c-11eb-83de-eaf82d512766.png)

![image](https://user-images.githubusercontent.com/3045767/116824837-97791c80-ab8c-11eb-9751-cd7b8ec5da8a.png)

**Comment by zakhenry** [comment]
Yea I had a dig around to see if there was any later ES2020 feature added that is more easily detected, but came up empty handed. It seems like a reasonable enough compromise to be able to support clients that ban unsafe js eval, which I would hope would be more than 0.4% of users!

**Comment by zakhenry** [comment]
I wonder since this repo also has a demo site whether it would be worth putting in a meta tag to ensure this doesn't regress?

```html
<!-- might need more policies, but not `script-src: 'unsafe-eval'`! -->
<meta http-equiv="Content-Security-Policy" content="default-src 'self';"> 
```

**Comment by stephanrauh** [comment]
Does that work with a simple Apache http server?

Side remark: I dream of moving both my blog and https://pdfviewer.net to the AWS, but I estimate that'd amount to a 100€ bill each. Possibly a lot more, if a hacker runs a DDOS attack. I'm sure I can get it a lot cheaper, but it's difficult to migrate back and forth, so I need to run a small-scale experiment first. Plus, I need to learn more about the features of the AWS.

**Comment by zakhenry** [comment]
Yea the `<meta>` bit should just go in the `<head>` of your `index.html`, I don't think Apache config could interfere.

Is there a good reason not to use github pages for pdfviewer.net ?  Does it do server side computation somewhere?

**Comment by stephanrauh** [comment]
Currently there's no server-side computation. But I'd like to add server-side stuff. Comments, visitor counter, and - high on my wish list! - server-side rendering.

**Comment by zakhenry** [comment]
Also while I admit very little knowledge of implementing such a thing, I think AWS could be surprisingly cheap if you went for a more serverless architecture - AWS Lambda allows 1 million requests free per month, and you could also put Cloudfront in front of a server as a CDN, which costs on the order of 10 cents per month for modest loads

**Comment by stephanrauh** [comment]
That sounds encouraging. On average, I see roughly 700.000 http requests with 50 GB traffic. I'm sure that includes several hacker attacks.

**Comment by stephanrauh** [comment]
Oh, whow, activating CSP in the showcase seems to require some work...

![image](https://user-images.githubusercontent.com/3045767/117205499-fe4b3f80-adf1-11eb-96b3-3d1e7b9524f7.png)

**Comment by stephanrauh** [comment]
I've just published version 9.0.0-alpha.5. I hope it solves your issue. Can you give it a try, please?

**Comment by zakhenry** [comment]
Hah yea turning on strict CSP can be a bit of an eye opener.

An yep, I can confirm beta 5 works perfectly for my application 🎉 thanks!
